demos/speculation_over_exec_hw_breakpoint.cc

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Demonstrates speculative execution over hardware breakpoint fault.
+ * That is a breakpoint that guards an instruction address and is triggered when
+ * that instruction is executed (not read nor written).
+ * We fork the process and run the demonstration in the child, while the parent
+ * takes care for setting up the breakpoint and moving the instruction pointer
+ * over the dead code after the fault.
+ **/
+
+#ifndef __linux__
+#  error Unsupported OS. Linux required.
+#endif
+
+#if !defined(__i386__) && !defined(__x86_64__)
+#  error Unsupported CPU. X86/64 required.
+#endif
+
+#include <array>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include <signal.h>
+#include <sys/ptrace.h>
+#include <sys/types.h>
+#include <sys/user.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+#include "cache_sidechannel.h"
+#include "instr.h"
+#include "utils.h"
+
+const char *public_data = "Hello, world!";
+const char *private_data = "It's a s3kr3t!!!";
+
+// Points to the "nop" instruction that will be guarded by the execution
+// breakpoint.
+extern char breakpoint[];
+
+static char LeakByte(const char *data, size_t offset) {
+  CacheSideChannel sidechannel;
+  const std::array<BigByte, 256> &oracle = sidechannel.GetOracle();
+
+  for (int run = 0;; ++run) {
+    size_t safe_offset = run % strlen(public_data);
+    sidechannel.FlushOracle();
+
+    // We have to precompute the addresses in the oracle, because here the
+    // speculation window on Intel (not on AMD) is too small to allow
+    // computation of the unsafe address in the oracle speculatively.
+    const void *safe_address =
+        oracle.data() + static_cast<size_t>(data[safe_offset]);
+
+    // Architecturally dead variable - never read again. It is also the only
+    // fetch of the "data[offset]". Therefore its value is architecturally
+    // isolated from the rest of the program.
+    const void *unsafe_address =
+        oracle.data() + static_cast<size_t>(data[offset]);
+
+    // Successful access of the safe address in the Oracle.
+    ForceRead(safe_address);
+
+    // NOP instruction after the breakpoint label. That one is guarded by the
+    // execution breakpoint. Contrary to the read/write hardware watcher, this
+    // is a fault (not a trap) and the tracer moves the instruction pointer to
+    // afterspeculation instead.
+    asm volatile(
```

Comment:
I really wanted to propose using [labels as values](https://gcc.gnu.org/onlinedocs/gcc/Labels-as-Values.html) here, since it's supported by all the compilers we're likely to care about on Linux (GCC, ICC, Clang). Unfortunately an experiment showed that trying to use the value for anything other than a computed goto (`goto *addr`) behaves poorly: https://godbolt.org/z/Qf5ym-

Reply:
OK. I see.